### PR TITLE
BUG: Categorical.__setitem__ allows for tuple assignment

### LINKED ISSUE #20439: Categorical.__setitem__ doesn't handle setting tuples
We allow tuples in the categories, since they're hashable. But `Categorical.__setitem__` has an incorrect check before setting.

```python
In [3]: c = pd.Categorical([(0, 1), (1, 2)])

In [5]: c[0] = (1, 2)
```

Raises

```pytb
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-5-4d3f2e8a3cfd> in <module>()
----> 1 c[0] = (1, 2)

/Users/taugspurger/miniconda3/envs/pandas-0.19.2/lib/python3.5/site-packages/pandas/core/categorical.py in __setitem__(self, key, value)
   1652         # something to np.nan
   1653         if len(to_add) and not isnull(to_add).all():
-> 1654             raise ValueError("Cannot setitem on a Categorical with a new "
   1655                              "category, set the categories first")
   1656

ValueError: Cannot setitem on a Categorical with a new category, set the categories first
```

If it's easy, this is worth supporting. If it's complicated, then maybe not.

### PR DESCRIPTION
**Pull request by alysivji**
- [x] closes #20439 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Pretty straightforward fix in the `__setitem__` method

While I was looking into this, I discovered a bug that does not allow for the creation of mixed-dtype Categoricals, if the first element is not a tuple. Should I create an issue?

```console
In [20]: s = pd.Categorical([('a', 'a'), ('a', 'b'), ('b', 'a'), 'c'])

In [21]: s
Out[21]:
[(a, a), (a, b), (b, a), c]
Categories (4, object): [(a, a), (a, b), (b, a), c]

In [22]: s = pd.Categorical(['c', ('a', 'b'), ('b', 'a'), 'c'])
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-22-0f4b5f338532> in <module>()
----> 1 s = pd.Categorical(['c', ('a', 'b'), ('b', 'a'), 'c'])

~/Documents/siv-dev/projects/open-source/pandas/pandas/core/arrays/categorical.py in __init__(self, values, categories, ordered, dtype, fastpath)
    328             # _sanitize_array coerces np.nan to a string under certain versions
    329             # of numpy
--> 330             values = maybe_infer_to_datetimelike(values, convert_dates=True)
    331             if not isinstance(values, np.ndarray):
    332                 values = _convert_to_list_like(values)

~/Documents/siv-dev/projects/open-source/pandas/pandas/core/dtypes/cast.py in maybe_infer_to_datetimelike(value, convert_dates)
    893     if not is_list_like(v):
    894         v = [v]
--> 895     v = np.array(v, copy=False)
    896
    897     # we only care about object dtypes

ValueError: setting an array element with a sequence
```

### CODE REVIEW

File: pandas/core/arrays/categorical.py [hunk 1]
```diff
@@ -1971,10 +1971,15 @@ def __setitem__(self, key, value):
                 raise ValueError("Cannot set a Categorical with another, "
                                  "without identical categories")
 
-        rvalue = value if is_list_like(value) else [value]
-
+        # is the new value one of the defined categories?
```
Comment (gfyoung) [inline pandas/core/arrays/categorical.py]: Remove this.

File: pandas/tests/categorical/test_indexing.py [hunk 2]
```diff
@@ -103,3 +103,21 @@ def f():
             s.categories = [1, 2]
 
         pytest.raises(ValueError, f)
+
+    def test_setitem_with_tuple_categories(self):
```
Comment (gfyoung) [inline pandas/tests/categorical/test_indexing.py]: Reference the issue as a comment.

**Comment by gfyoung** [comment]
> Should I create an issue?

@alysivji : Yes, by all means!

**Review by gfyoung — APPROVED** [review]
LGTM!

cc @jreback

**Comment by alysivji** [comment]
With the 0.23.1 release last night, moved my whatsnew entry to 0.23.2

File: pandas/core/arrays/categorical.py [hunk 3]
```diff
@@ -1971,10 +1971,14 @@ def __setitem__(self, key, value):
                 raise ValueError("Cannot set a Categorical with another, "
                                  "without identical categories")
 
-        rvalue = value if is_list_like(value) else [value]
-
         from pandas import Index
-        to_add = Index(rvalue).difference(self.categories)
+        if isinstance(value, tuple):
+            rvalue = [value]
```
Comment (jreback) [inline pandas/core/arrays/categorical.py]: so the big problem I have here is that this should be a TypeError if the categories are not actual tuples. We dont' have an easy / reliable way to check this as its just an object Index. This is adding more and more technical debt on top of tupleize_cols.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/categorical.py (hunk 3 above)
Comment (alysivji) [reply pandas/core/arrays/categorical.py]: I'm gonna look into this more over the weekend / next week.

Thinking of sets, but want to understand `tupleize_cols` before I go another way.

**Pull request closed without merging**